### PR TITLE
Set a static Retry-After when evicting a pod is not allowed.

### LINKED ISSUE #36409: TooManyRequests(429) should have a non-empty RetryAfter when evicting a pod is not allowed
This is a FEATURE REQUEST.

AFAIK, eviciton is first used in `kubectl drain` in #35483.
If a eviction is not allowed, the caller will get a `429 Too Many Requests`, without a RetryAfter advice. See, [this](http://kubernetes.io/docs/admin/disruptions/#requesting-an-eviction)
It looks like this:
{
  "metadata": {},
  "status": "Failure",
  "message": "Cannot evict pod as it would violate the pod's disruption budget.",
  "code": 429
}

It will be very helpful for `kubectl drain` and others to have a non-empty `RetryAfter` field.

cc: @caesarxuchao

### PR DESCRIPTION
**Pull request by josephw**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

With the errors added in d3be1ac92eb64 it's easy to implement part of the TODO added way back in c86b84ccc0b, and address this improvement request.

It stops short of #84903.
 
**Which issue(s) this PR fixes**:

Fixes #36409.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by jtslear** [comment]
Finding a home, if the assignment is incorrect, please assist in reassessing and reassigning as necessary:

/sig api-machinery

**Comment by sftim** [comment]
I think also relevant to
/sig scheduling

**Comment by ahg-g** [comment]
/remove-sig scheduling

the scheduler doesn't invoke this code.

**Comment by josephw** [comment]
Hey @deads2k. I've rebased this to resolve conflicts. [Earlier comment](https://github.com/kubernetes/kubernetes/pull/90299#issuecomment-616580084) suggests this needs your approval; any feedback?

**Comment by josephw** [comment]
@justinsb, @wojtek-t; what can do to make progress with this?

File: pkg/registry/core/pod/storage/eviction.go [hunk 1]
```diff
@@ -303,11 +303,11 @@ func resourceVersionIsUnset(options *metav1.DeleteOptions) bool {
...
 
 func createTooManyRequestsError(name string) error {
-	// TODO(mml): Add a Retry-After header.  Once there are time-based
+	// TODO(mml): Once there are time-based
 	// budgets, we can sometimes compute a sensible suggested value.  But
-	// even without that, we can give a suggestion (10 minutes?) that
+	// even without that, we can give a suggestion (10 minutes) that
 	// prevents well-behaved clients from hammering us.
-	err := errors.NewTooManyRequests("Cannot evict pod as it would violate the pod's disruption budget.", 0)
+	err := errors.NewTooManyRequests("Cannot evict pod as it would violate the pod's disruption budget.", 600)
```
Comment (wojtek-t) [inline pkg/registry/core/pod/storage/eviction.go:310]: Looking into the code, it seems that this path isn't really exercised on overload, but rather in two options:
(a) on conflict to force the client to retry:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L262-L265
- if we're lagging behind the current version:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L319

For (a) - returning 429 seems strange to me - what's the purpose of hiding "Confict" here? Users should retry conflict too? I think returning 429 is very misleading in this case.


For (b) - returning 429 is also misleading. I'm not sure what would be the option though - maybe 425 (StatusTooEarly):
https://golang.org/src/net/http/status.go

@liggitt - WDYT?

File: pkg/registry/core/pod/storage/eviction.go [hunk 2]
```diff
@@ -303,11 +303,11 @@ func resourceVersionIsUnset(options *metav1.DeleteOptions) bool {
 }
 
 func createTooManyRequestsError(name string) error {
-	// TODO(mml): Add a Retry-After header.  Once there are time-based
+	// TODO(mml): Once there are time-based
```
Comment (liggitt) [inline pkg/registry/core/pod/storage/eviction.go:306]: suggest switching to an active user

File: pkg/registry/core/pod/storage/eviction.go (hunk 1 above)
Comment (liggitt) [reply pkg/registry/core/pod/storage/eviction.go:310]: The server already retries on conflict. This is surfaced to the client if that retry was exhausted.

If the client submits an eviction without any preconditions, it isn't really expecting a conflict response.

A 429 doesn't seem terrible here... it means something is overloading the server with PDB updates... either pod churn, or eviction churn.

Comment (liggitt) [reply pkg/registry/core/pod/storage/eviction.go:310]: that said, a 10-minute backoff seems excessive to me

File: pkg/registry/core/pod/storage/eviction.go (hunk 2 above)
Comment (josephw) [reply pkg/registry/core/pod/storage/eviction.go:306]: Without a clear candidate, is it better to remove the username, or leave it as-is?

**Comment by josephw** [comment]
> that said, a 10-minute backoff seems excessive to me

That's the [original value suggested](https://github.com/kubernetes/kubernetes/commit/c86b84ccc0b318c158c6ee963ea723efa40a5ef3#diff-59770c9504852cc1abd9f5783358c05a37b64337090284c2f13167d65258ee2cR178); would two minutes be a better initial choice today?

**Comment by wojtek-t** [comment]
> That's the original value suggested; would two minutes be a better initial choice today?

Even 2 minutes seem like too high to me.
As for the "stale cache" example - I've seen this stale by few seconds, but almost never by more than 10s.

So I would suggest setting to something like 5-10s.
If it appears to be too small, we can always increase this, but even that will be a clear improvement comparing to what we have now.

**Comment by wojtek-t** [comment]
@josephw - please squash commits and I will approve

**Pull request merged**